File: apk/app/src/main/java/com/example/momentra/ui/shell/shared/MomentCurrencyResolver.kt
package com.example.momentra.ui.shell.shared

import com.example.momentra.data.api.BusinessFinanceTotalDto
import com.example.momentra.data.api.GroupFinanceTotalDto
import com.example.momentra.data.api.GroupSetupBudgetDto
import com.example.momentra.ui.shell.group.shared.GroupFinanceFormat
import com.example.momentra.ui.shell.group.shared.TravelCurrencyCatalog
import java.math.BigDecimal

/**
 * Setup-aware currency resolution for money sheets.
 * Callers fetch finance / setup data, then pass it here — no network I/O.
 */
object MomentCurrencyResolver {
    data class FinanceTotal(
        val currencyCode: String,
        val budgetTotal: String? = null,
        val expenseTotal: String? = null,
    )

    data class SetupBudget(
        val currencyCode: String,
        val isPrimary: Boolean = false,
    )

    fun fromGroupTotals(totals: List<GroupFinanceTotalDto>): List<FinanceTotal> =
        totals.map { FinanceTotal(it.currencyCode, it.budgetTotal, it.expenseTotal) }

    fun fromBusinessTotals(totals: List<BusinessFinanceTotalDto>): List<FinanceTotal> =
        totals.map { FinanceTotal(it.currencyCode, expenseTotal = it.expenseTotal) }

    fun fromGroupSetupBudgets(budgets: List<GroupSetupBudgetDto>): List<SetupBudget> =
        budgets.map { SetupBudget(it.currencyCode, it.isPrimary == true) }

    fun resolveMomentCurrency(
        financeTotals: List<FinanceTotal> = emptyList(),
        setupBudgets: List<SetupBudget> = emptyList(),
        setupPreferences: Map<String, Any?>? = null,
        accountCurrency: String? = null,
    ): String {
        primaryFinanceTotal(financeTotals)?.currencyCode?.let { normalize(it) }?.let { return it }
        primarySetupBudget(setupBudgets)?.currencyCode?.let { normalize(it) }?.let { return it }
        preferenceCurrency(setupPreferences)?.let { return it }
        accountCurrency?.let { normalize(it) }?.let { return it }
        return "INR"
    }

    fun resolveMomentCurrencies(
        financeTotals: List<FinanceTotal> = emptyList(),
        setupBudgets: List<SetupBudget> = emptyList(),
        setupPreferences: Map<String, Any?>? = null,
        accountCurrency: String? = null,
    ): List<String> {
        val out = mutableListOf<String>()
        fun append(code: String?) {
            val normalized = code?.let { normalize(it) } ?: return
            if (normalized !in out) out.add(normalized)
        }

        financeTotals
            .sortedByDescending { financeWeight(it) }
            .forEach { append(it.currencyCode) }

        val primaryBudget = primarySetupBudget(setupBudgets)
        append(primaryBudget?.currencyCode)
        setupBudgets
            .filter { it != primaryBudget }
            .forEach { append(it.currencyCode) }

        append(preferenceCurrency(setupPreferences))
        extraCurrencies(setupPreferences).forEach { append(it) }
        append(accountCurrency?.let { normalize(it) })

        return out
    }

    fun pickerOptions(preferred: List<String>): List<String> {
        val preferredNorm = preferred.mapNotNull { normalize(it) }.distinct()
        val rest = TravelCurrencyCatalog.codes.filter { it !in preferredNorm }
        return preferredNorm + rest
    }

    private fun primaryFinanceTotal(totals: List<FinanceTotal>): FinanceTotal? {
        if (totals.isEmpty()) return null
        return totals.maxByOrNull { financeWeight(it) }
    }

    private fun primarySetupBudget(budgets: List<SetupBudget>): SetupBudget? {
        if (budgets.isEmpty()) return null
        return budgets.firstOrNull { it.isPrimary } ?: budgets.first()
    }

    private fun financeWeight(total: FinanceTotal): Int {
        if (GroupFinanceFormat.parseAmount(total.budgetTotal) > BigDecimal.ZERO) return 2
        if (GroupFinanceFormat.parseAmount(total.expenseTotal) > BigDecimal.ZERO) return 1
        return 0
    }

    private fun preferenceCurrency(prefs: Map<String, Any?>?): String? =
        normalize(prefs?.get("currency")?.toString())

    private fun extraCurrencies(prefs: Map<String, Any?>?): List<String> {
        val raw = prefs?.get("extraCurrencies") ?: return emptyList()
        return when (raw) {
            is List<*> -> raw.mapNotNull { normalize(it?.toString()) }
            is Array<*> -> raw.mapNotNull { normalize(it?.toString()) }
            else -> emptyList()
        }
    }

    private fun normalize(code: String?): String? {
        val upper = code?.trim()?.uppercase()?.take(3) ?: return null
        if (upper.length != 3) return null
        return if (TravelCurrencyCatalog.find(upper) != null) upper else upper
    }
}

File: momentra/momentra/Shell/Shared/MomentCurrencyResolver.swift
import Foundation

/// Setup-aware currency resolution for money sheets.
/// Callers fetch finance / setup data, then pass it here — no network I/O.
enum MomentCurrencyResolver {
    struct FinanceTotal: Equatable {
        let currencyCode: String
        var budgetTotal: String? = nil
        var expenseTotal: String? = nil
    }

    struct SetupBudget: Equatable {
        let currencyCode: String
        var isPrimary: Bool = false
    }

    static func fromGroupTotals(_ totals: [APIClient.GroupFinanceTotalsPayload]) -> [FinanceTotal] {
        totals.map { FinanceTotal(currencyCode: $0.currencyCode, budgetTotal: $0.budgetTotal, expenseTotal: $0.expenseTotal) }
    }

    static func fromBusinessTotals(_ totals: [APIClient.BusinessFinanceTotalsPayload]) -> [FinanceTotal] {
        totals.map { FinanceTotal(currencyCode: $0.currencyCode, expenseTotal: $0.expenseTotal) }
    }

    static func fromGroupSetupBudgets(_ budgets: [GroupSetupBudgetPrefill]) -> [SetupBudget] {
        budgets.map { SetupBudget(currencyCode: $0.currencyCode, isPrimary: $0.isPrimary == true) }
    }

    static func resolveMomentCurrency(
        financeTotals: [FinanceTotal] = [],
        setupBudgets: [SetupBudget] = [],
        setupPreferences: [String: Any]? = nil,
        accountCurrency: String? = nil
    ) -> String {
        if let code = normalize(primaryFinanceTotal(financeTotals)?.currencyCode) { return code }
        if let code = normalize(primarySetupBudget(setupBudgets)?.currencyCode) { return code }
        if let code = preferenceCurrency(setupPreferences) { return code }
        if let code = accountCurrency.flatMap({ normalize($0) }) { return code }
        return "INR"
    }

    static func resolveMomentCurrencies(
        financeTotals: [FinanceTotal] = [],
        setupBudgets: [SetupBudget] = [],
        setupPreferences: [String: Any]? = nil,
        accountCurrency: String? = nil
    ) -> [String] {
        var out: [String] = []
        func append(_ code: String?) {
            guard let normalized = code.flatMap({ normalize($0) }) else { return }
            if !out.contains(normalized) { out.append(normalized) }
        }

        financeTotals
            .sorted { financeWeight($0) > financeWeight($1) }
            .forEach { append($0.currencyCode) }

        let primaryBudget = primarySetupBudget(setupBudgets)
        append(primaryBudget?.currencyCode)
        setupBudgets
            .filter { $0.currencyCode != primaryBudget?.currencyCode }
            .forEach { append($0.currencyCode) }

        append(preferenceCurrency(setupPreferences))
        extraCurrencies(setupPreferences).forEach { append($0) }
        append(accountCurrency.flatMap { normalize($0) })

        return out
    }

    static func pickerOptions(preferred: [String]) -> [String] {
        let preferredNorm = preferred.compactMap { normalize($0) }
        var seen = Set<String>()
        let orderedPreferred = preferredNorm.filter { seen.insert($0).inserted }
        let rest = TravelCurrencyCatalog.codes.filter { !seen.contains($0) }
        return orderedPreferred + rest
    }

    private static func primaryFinanceTotal(_ totals: [FinanceTotal]) -> FinanceTotal? {
        totals.max { financeWeight($0) < financeWeight($1) }
    }

    private static func primarySetupBudget(_ budgets: [SetupBudget]) -> SetupBudget? {
        guard !budgets.isEmpty else { return nil }
        return budgets.first(where: \.isPrimary) ?? budgets.first
    }

    private static func financeWeight(_ total: FinanceTotal) -> Int {
        if GroupFinanceFormat.parseAmount(total.budgetTotal) > 0 { return 2 }
        if GroupFinanceFormat.parseAmount(total.expenseTotal) > 0 { return 1 }
        return 0
    }

    private static func preferenceCurrency(_ prefs: [String: Any]?) -> String? {
        guard let raw = prefs?["currency"] else { return nil }
        if let s = raw as? String { return normalize(s) }
        return normalize(String(describing: raw))
    }

    private static func extraCurrencies(_ prefs: [String: Any]?) -> [String] {
        guard let raw = prefs?["extraCurrencies"] else { return [] }
        if let list = raw as? [String] { return list.compactMap { normalize($0) } }
        if let list = raw as? [Any] { return list.compactMap { normalize(String(describing: $0)) } }
        return []
    }

    private static func normalize(_ code: String?) -> String? {
        guard let trimmed = code?.trimmingCharacters(in: .whitespacesAndNewlines).uppercased(),
              trimmed.count == 3 else { return nil }
        return trimmed
    }
}
